blockchain/message_pool/src/msgpool.rs

```diff
@@ -0,0 +1,1066 @@
+    }
+
+    api.put_message(&msg)?;
+    let msett = pending.get_mut(msg.message().from());
+    match msett {
+        Some(mset) => mset.add(&msg)?,
+        None => {
+            let mut mset = MsgSet::new();
+            mset.add(&msg)?;
+            pending.insert(*msg.message().from(), mset);
+        }
+    }
+    Ok(())
+}
+
+/// This function will revert and/or apply tipsets to the message pool. This function should be
+/// called every time that there is a head change in the message pool
+pub async fn head_change<T>(
+    api: Arc<RwLock<T>>,
+    bls_sig_cache: Arc<RwLock<LruCache<Cid, Signature>>>,
+    pending: Arc<RwLock<HashMap<Address, MsgSet>>>,
+    cur_tipset: Arc<RwLock<Tipset>>,
+    revert: Vec<Tipset>,
+    apply: Vec<Tipset>,
+) -> Result<(), Error>
+where
+    T: Provider,
+{
+    let mut rmsgs: HashMap<Address, HashMap<u64, SignedMessage>> = HashMap::new();
+    for ts in revert {
+        let api_locked = api.write().await;
+        let pts = api_locked.load_tipset(ts.parents())?;
+        drop(api_locked);
+
+        let mut msgs: Vec<SignedMessage> = Vec::new();
+        for block in ts.blocks() {
+            let api_locked = api.read().await;
+            let (umsg, mut smsgs) = api_locked.messages_for_block(&block)?;
+            drop(api_locked);
+            msgs.append(smsgs.as_mut());
+            for msg in umsg {
+                let bls_sig_cache = bls_sig_cache.clone();
+                let smsg = recover_sig(bls_sig_cache, msg).await?;
+                msgs.push(smsg)
+            }
+        }
+        let parent = pts.clone();
+        let mut cur_ts_locked = cur_tipset.write().await;
+        *cur_ts_locked = parent;
+        drop(cur_ts_locked);
+
+        for msg in msgs {
+            add(msg, rmsgs.borrow_mut());
+        }
+    }
+
+    for ts in apply {
+        for b in ts.blocks() {
+            let api_locked = api.read().await;
+            let (msgs, smsgs) = api_locked.messages_for_block(b)?;
+            drop(api_locked);
+
+            for msg in smsgs {
+                rm(
+                    msg.from(),
+                    pending.clone(),
+                    msg.sequence(),
+                    rmsgs.borrow_mut(),
+                )
+                .await;
+            }
+            for msg in msgs {
+                rm(
+                    msg.from(),
+                    pending.clone(),
+                    msg.sequence(),
+                    rmsgs.borrow_mut(),
+                )
+                .await;
+            }
+        }
+        let mut cur_ts_locked = cur_tipset.write().await;
+        *cur_ts_locked = ts;
+        drop(cur_ts_locked);
+    }
+    for (_, hm) in rmsgs {
+        for (_, msg) in hm {
+            let api_locked = api.clone();
+            let pending = pending.clone();
+            let bls_sig_cache = bls_sig_cache.clone();
+            add_helper(api_locked, bls_sig_cache, pending, msg.clone())
+                .await
+                .ok();
```

Comment:
```suggestion
            if let Err(e) =
                add_helper(api.clone(), bls_sig_cache.clone(), pending.clone(), msg).await
            {
                error!("Failed to readd message from reorg to mpool: {}", e);
            }
```
rn you are ignoring that error, should log it (error macro is just log::error)